src/App.tsx

```diff
@@ -1,24 +1,208 @@
+const deleteTodo = (todoId: number) => client.delete(`/todos/${todoId}`);
+
+const prepareTodos = (todoList: Todo[], filterType: FilterType) => (
+  todoList.filter(todo => {
+    switch (filterType) {
+      case FilterType.ACTIVE:
+        return !todo.completed;
+
+      case FilterType.COMPLETED:
+        return todo.completed;
+
+      default:
+        return true;
+    }
+  })
+);
+
+const getActiveTodosCount = (todoList: Todo[]) => (
+  todoList.filter(todo => !todo.completed).length
+);
 
 export const App: React.FC = () => {
+  const [todoList, setTodoList] = useState<Todo[] | null>(null);
+  const [filterType, setFilterType] = useState(FilterType.ALL);
+  const [errorType, setErrorType] = useState(ErrorType.NONE);
+  const [isErrorShown, setIsErrorShown] = useState(false);
+  const [todoInputValue, setTodoInputValue] = useState('');
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [isAddDisabled, setIsAddDisabled] = useState(false);
+  const [deletingId, setDeletingId] = useState<number | null>(null);
+
+  const getTodoList = useCallback(async () => {
+    const todos = await getTodos();
+
+    setTodoList(todos);
+  }, []);
+
+  const preparedTodos = useMemo(() => (
+    prepareTodos(todoList || [], filterType)
+  ), [todoList, filterType]);
+
+  const activeTodosCount = useMemo(() => (
+    getActiveTodosCount(todoList || [])
+  ), [todoList]);
+
+  const areCompletedTodos = todoList
+    ? activeTodosCount < todoList.length
+    : false;
+
+  const handleFilterChange = (newFilterType: FilterType) => {
+    setFilterType(newFilterType);
+  };
+
+  const handleTodoInputChange = (value: string) => {
+    setIsErrorShown(false);
+
+    setTodoInputValue(value);
+  };
+
+  const executePostTodo = () => {
+    const addedTodo = {
+      title: todoInputValue,
+      userId: USER_ID,
+      completed: false,
+    };
+
+    setTempTodo({
+      id: 0,
+      ...addedTodo,
+    });
+
+    postTodo(addedTodo)
+      .then(response => {
+        const {
+          id,
+          title,
+          userId,
+          completed,
+        } = response;
+        const newTodo = {
+          id,
+          title,
+          userId,
+          completed,
+        };
+
+        setTodoList(currentList => (
+          currentList
+            ? [...currentList, newTodo]
+            : [newTodo]
+        ));
+      })
+      .catch(() => {
+        setErrorType(ErrorType.ADD);
+        setIsErrorShown(true);
+        throw new Error('Add todo error');
+      })
+      .finally(() => {
+        setTempTodo(null);
+      });
+  };
+
+  const handleAddTodo = async (event: FormEvent<HTMLFormElement>) => {
+    event.preventDefault();
+
+    if (todoInputValue.trim().length === 0) {
+      setIsErrorShown(true);
+      setErrorType(ErrorType.TITLE);
+
+      return;
+    }
+
+    setIsAddDisabled(true);
+    executePostTodo();
+    setIsAddDisabled(false);
```

Comment:
why you need to do this?
It can update in the same second

Reply:
We have a directive to disable the input until receiving a response from the API. But without "await" it didn't work.